File: project-todo-app/src/app/services/todo.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class TodoService {
  private baseUrl = 'http://localhost:8082/api/projects';
  private todoUrl = 'http://localhost:8082/api/todos';
  constructor(private http: HttpClient) { }

  addTodo(todo: any): Observable<any> {
    return this.http.post(`${this.todoUrl}`, todo);
  }

  getTodos(pid: any): Observable<any> {
    return this.http.get(`${this.baseUrl}/` + pid);
  }
  getTodoById(pid: any, tid: any): Observable<any> {
    return this.http.get(`${this.todoUrl}/${pid}/todos/${tid}`);
  }


  markTodoAsComplete(pid:any,todoId: number): Observable<any> {
    return this.http.put(`${this.todoUrl}/${pid}/complete/${todoId}`, {});
  }

  markTodoAsPending(pid:any,todoId: number): Observable<any> {
    return this.http.put(`${this.todoUrl}/${pid}/pending/${todoId}`, {});
  }

  deleteTodo(pid:any,todoId: number): Observable<any> {
    return this.http.delete(`${this.todoUrl}/${pid}/todos/${todoId}`);
  }
  updateTodo(pid: any, tid: any, data: any) {
    console.log(pid);
    
    return this.http.put(`${this.todoUrl}/${pid}/todos/${tid}`, data);
  }
}

File: Project-Management/src/main/java/com/example/demo/service/TodoService.java
package com.example.demo.service;

import com.example.demo.dto.ResponseDto;
import com.example.demo.dto.TodoDto;
import org.springframework.http.ResponseEntity;

public interface TodoService {

	ResponseEntity<ResponseDto> addTodo(TodoDto todoDto);

	ResponseEntity<ResponseDto> updateTodo(TodoDto updatedTodoDto);

	ResponseEntity<ResponseDto> deleteTodo(Integer pid, Integer tid);

	ResponseEntity<ResponseDto> markTodoAsComplete(Integer pid, Integer id);

	public ResponseEntity<ResponseDto> updateTodo(Integer pid, Integer tid, TodoDto updatedTodoDto);

	public ResponseEntity<ResponseDto> markTodoAsPending(Integer pid, Integer id);

	public ResponseEntity<ResponseDto> getTodoById(Integer pid, Integer tid);
}
